File: agents/Group37/heuristics.py
"""Hex-specific heuristics for MCTS enhancement."""

import heapq
import random as _random

from src.Board import Board
from src.Colour import Colour
from src.Tile import Tile


# Two-bridge patterns: relative offsets that form virtual connections
# A two-bridge exists when two stones of same color can be connected
# via two empty cells that share neighbors with both stones
TWO_BRIDGE_PATTERNS = [
    # (dx1, dy1, dx2, dy2, empty1, empty2) - stone offsets and empty cell offsets
    ((1, 0), (0, 1), (1, 1)),     # Bridge pattern 1
    ((-1, 1), (0, 1), (-1, 2)),   # Bridge pattern 2
    ((-1, 0), (-1, 1), (-2, 1)),  # Bridge pattern 3
    ((1, -1), (0, -1), (1, -2)),  # Bridge pattern 4
    ((1, 0), (1, -1), (2, -1)),   # Bridge pattern 5
    ((-1, 0), (0, -1), (-1, -1)), # Bridge pattern 6
]


def get_neighbors(x: int, y: int, size: int) -> list[tuple[int, int]]:
    """Get valid neighboring positions."""
    neighbors = []
    for k in range(Tile.NEIGHBOUR_COUNT):
        nx = x + Tile.I_DISPLACEMENTS[k]
        ny = y + Tile.J_DISPLACEMENTS[k]
        if 0 <= nx < size and 0 <= ny < size:
            neighbors.append((nx, ny))
    return neighbors


def detect_virtual_connections(
    board: Board, colour: Colour
) -> list[tuple[int, int]]:
    """
    Detect moves that complete virtual connections (two-bridges).
    Returns list of moves that would complete a two-bridge pattern.
    """
    size = board.size
    completing_moves = []

    for x in range(size):
        for y in range(size):
            if board.tiles[x][y].colour != colour:
                continue

            # Check each two-bridge pattern
            for pattern in TWO_BRIDGE_PATTERNS:
                (dx1, dy1), (dx2, dy2), (ex, ey) = pattern

                # Position of the other stone in the bridge
                x2, y2 = x + dx1 + dx2, y + dy1 + dy2

                if not (0 <= x2 < size and 0 <= y2 < size):
                    continue
                if board.tiles[x2][y2].colour != colour:
                    continue

                # Check the two empty cells that form the bridge
                e1x, e1y = x + dx1, y + dy1
                e2x, e2y = x + ex, y + ey

                if not (0 <= e1x < size and 0 <= e1y < size):
                    continue
                if not (0 <= e2x < size and 0 <= e2y < size):
                    continue

                e1_empty = board.tiles[e1x][e1y].colour is None
                e2_empty = board.tiles[e2x][e2y].colour is None

                # If one is occupied by opponent, the completing move is important
                if e1_empty and not e2_empty:
                    if board.tiles[e2x][e2y].colour == Colour.opposite(colour):
                        completing_moves.append((e1x, e1y))
                elif e2_empty and not e1_empty:
                    if board.tiles[e1x][e1y].colour == Colour.opposite(colour):
                        completing_moves.append((e2x, e2y))

    return completing_moves


def shortest_path_distance(
    board: Board, colour: Colour
) -> int:
    """
    Calculate shortest path distance to victory using Dijkstra.
    Empty cells cost 1, own cells cost 0, opponent cells cost infinity.
    Returns distance (lower is better for the player).
    """
    size = board.size
    INF = size * size + 1

    # For RED: top to bottom (row 0 to row size-1)
    # For BLUE: left to right (col 0 to col size-1)

    dist = [[INF] * size for _ in range(size)]
    heap = []

    if colour == Colour.RED:
        # Start from top row
        for j in range(size):
            tile_colour = board.tiles[0][j].colour
            if tile_colour == Colour.opposite(colour):
                continue
            cost = 0 if tile_colour == colour else 1
            dist[0][j] = cost
            heapq.heappush(heap, (cost, 0, j))
    else:  # BLUE
        # Start from left column
        for i in range(size):
            tile_colour = board.tiles[i][0].colour
            if tile_colour == Colour.opposite(colour):
                continue
            cost = 0 if tile_colour == colour else 1
            dist[i][0] = cost
            heapq.heappush(heap, (cost, i, 0))

    while heap:
        d, x, y = heapq.heappop(heap)

        if d > dist[x][y]:
            continue

        # Check goal
        if colour == Colour.RED and x == size - 1:
            return d
        if colour == Colour.BLUE and y == size - 1:
            return d

        for nx, ny in get_neighbors(x, y, size):
            tile_colour = board.tiles[nx][ny].colour
            if tile_colour == Colour.opposite(colour):
                continue

            edge_cost = 0 if tile_colour == colour else 1
            new_dist = d + edge_cost

            if new_dist < dist[nx][ny]:
                dist[nx][ny] = new_dist
                heapq.heappush(heap, (new_dist, nx, ny))

    # Find minimum distance to goal edge
    min_dist = INF
    if colour == Colour.RED:
        for j in range(size):
            min_dist = min(min_dist, dist[size - 1][j])
    else:
        for i in range(size):
            min_dist = min(min_dist, dist[i][size - 1])

    return min_dist


def evaluate_position(board: Board, colour: Colour) -> float:
    """
    Evaluate board position for a player.
    Returns positive score if position is good for colour.
    """
    my_dist = shortest_path_distance(board, colour)
    opp_dist = shortest_path_distance(board, Colour.opposite(colour))

    # Lower distance is better, so we want opp_dist - my_dist
    return opp_dist - my_dist


def get_move_priority(
    board: Board,
    move: tuple[int, int],
    colour: Colour,
) -> float:
    """
    Calculate priority score for a move during simulation.
    Higher score = more likely to be selected.
    """
    x, y = move
    size = board.size
    score = 0.0

    # Bonus for extending existing connections
    for nx, ny in get_neighbors(x, y, size):
        if board.tiles[nx][ny].colour == colour:
            score += 2.0

    # Bonus for blocking opponent connections
    opp = Colour.opposite(colour)
    for nx, ny in get_neighbors(x, y, size):
        if board.tiles[nx][ny].colour == opp:
            score += 1.5

    # Bonus for center positions (strategic value)
    center = size // 2
    dist_to_center = abs(x - center) + abs(y - center)
    score += max(0, (size - dist_to_center) / size)

    # Bonus for positions along main diagonal (often strategically important)
    if colour == Colour.RED:
        # RED benefits from positions along y axis progression
        score += 0.5 if x == y else 0
    else:
        # BLUE benefits from positions along x axis progression
        score += 0.5 if x == y else 0

    return score


def select_weighted_move(
    empty_tiles: list[tuple[int, int]],
    board: Board,
    colour: Colour,
) -> tuple[int, int]:
    """
    Select a move with probability weighted by heuristic scores.
    Uses a simplified selection for speed.
    """
    if not empty_tiles:
        return None

    if len(empty_tiles) == 1:
        return empty_tiles[0]

    # For speed, only score a subset of moves
    sample_size = min(8, len(empty_tiles))

    # Get priorities for sampled moves
    import random
    if len(empty_tiles) <= sample_size:
        candidates = empty_tiles
    else:
        candidates = random.sample(empty_tiles, sample_size)

    best_move = candidates[0]
    best_score = -float('inf')

    for move in candidates:
        score = get_move_priority(board, move, colour)
        # Add small random noise to break ties
        score += random.random() * 0.1
        if score > best_score:
            best_score = score
            best_move = move

    return best_move


# Strong opening moves for swap rule evaluation
# These are first moves that are generally considered strong
STRONG_OPENINGS = {
    # Center and near-center positions
    (5, 5), (4, 5), (5, 4), (4, 6), (6, 4), (5, 6), (6, 5),
    (4, 4), (6, 6), (3, 5), (5, 3), (3, 6), (6, 3),
}

# Weak opening moves (edges and corners)
WEAK_OPENINGS = {
    (0, 0), (0, 10), (10, 0), (10, 10),
    (0, 1), (1, 0), (0, 9), (9, 0), (1, 10), (10, 1), (9, 10), (10, 9),
}


def should_swap(opp_move_x: int, opp_move_y: int, board_size: int = 11) -> bool:
    """
    Determine if we should swap based on opponent's opening move.
    Swap if the opening move is strong (we want that position).
    """
    move = (opp_move_x, opp_move_y)

    # Always swap strong openings
    if move in STRONG_OPENINGS:
        return True

    # Never swap weak openings
    if move in WEAK_OPENINGS:
        return False

    # For other moves, use distance from center heuristic
    center = board_size // 2
    dist = abs(opp_move_x - center) + abs(opp_move_y - center)

    # Swap if within 2 tiles of center
    return dist <= 2


# Zobrist hashing for transposition table
_random.seed(42)  # Deterministic for reproducibility

ZOBRIST_TABLE: dict[tuple[int, int, Colour], int] = {}

def _init_zobrist(size: int = 11):
    """Initialize Zobrist hash table."""
    global ZOBRIST_TABLE
    if ZOBRIST_TABLE:
        return
    for i in range(size):
        for j in range(size):
            for colour in [Colour.RED, Colour.BLUE]:
                ZOBRIST_TABLE[(i, j, colour)] = _random.getrandbits(64)

_init_zobrist()


def compute_board_hash(board: Board) -> int:
    """Compute Zobrist hash of board state."""
    h = 0
    for i in range(board.size):
        for j in range(board.size):
            colour = board.tiles[i][j].colour
            if colour is not None:
                h ^= ZOBRIST_TABLE.get((i, j, colour), 0)
    return h


def update_hash(current_hash: int, x: int, y: int, colour: Colour) -> int:
    """Incrementally update hash after a move."""
    return current_hash ^ ZOBRIST_TABLE.get((x, y, colour), 0)

File: agents/Group37/cpp/heuristics.cpp
#include "heuristics.h"
#include <algorithm>
#include <cmath>
#include <random>
#include <queue>
#include <limits>
#include <unordered_map>

namespace hex {

// Strong openings (center and near-center positions)
// From Python: (5, 5), (4, 5), (5, 4), (4, 6), (6, 4), (5, 6), (6, 5),
//              (4, 4), (6, 6), (3, 5), (5, 3), (3, 6), (6, 3)
const std::unordered_set<int> STRONG_OPENINGS = {
    position_to_key(5, 5), position_to_key(4, 5), position_to_key(5, 4),
    position_to_key(4, 6), position_to_key(6, 4), position_to_key(5, 6),
    position_to_key(6, 5), position_to_key(4, 4), position_to_key(6, 6),
    position_to_key(3, 5), position_to_key(5, 3), position_to_key(3, 6),
    position_to_key(6, 3)
};

// Weak openings (edges and corners)
// From Python: (0, 0), (0, 10), (10, 0), (10, 10),
//              (0, 1), (1, 0), (0, 9), (9, 0), (1, 10), (10, 1), (9, 10), (10, 9)
const std::unordered_set<int> WEAK_OPENINGS = {
    position_to_key(0, 0), position_to_key(0, 10), position_to_key(10, 0), position_to_key(10, 10),
    position_to_key(0, 1), position_to_key(1, 0), position_to_key(0, 9), position_to_key(9, 0),
    position_to_key(1, 10), position_to_key(10, 1), position_to_key(9, 10), position_to_key(10, 9)
};

// Zobrist hash table
static std::unordered_map<int, uint64_t> ZOBRIST_TABLE;
static bool zobrist_initialized = false;

void init_zobrist() {
    if (zobrist_initialized) return;

    std::mt19937_64 rng(42); // Deterministic seed for reproducibility
    std::uniform_int_distribution<uint64_t> dist;

    for (int i = 0; i < BOARD_SIZE; ++i) {
        for (int j = 0; j < BOARD_SIZE; ++j) {
            for (int c = 1; c <= 2; ++c) { // RED=1, BLUE=2
                int key = i * BOARD_SIZE * 3 + j * 3 + c;
                ZOBRIST_TABLE[key] = dist(rng);
            }
        }
    }

    zobrist_initialized = true;
}

uint64_t compute_board_hash(const Board& board) {
    if (!zobrist_initialized) init_zobrist();

    uint64_t h = 0;
    for (int i = 0; i < BOARD_SIZE; ++i) {
        for (int j = 0; j < BOARD_SIZE; ++j) {
            Tile tile = board.get_tile(i, j);
            if (tile != Tile::EMPTY) {
                int colour_val = static_cast<int>(tile);
                int key = i * BOARD_SIZE * 3 + j * 3 + colour_val;
                h ^= ZOBRIST_TABLE[key];
            }
        }
    }
    return h;
}

uint64_t update_hash(uint64_t current_hash, int x, int y, Colour colour) {
    if (!zobrist_initialized) init_zobrist();

    int colour_val = static_cast<int>(colour);
    int key = x * BOARD_SIZE * 3 + y * 3 + colour_val;
    return current_hash ^ ZOBRIST_TABLE[key];
}

bool should_swap(int opp_move_x, int opp_move_y, int board_size) {
    int key = position_to_key(opp_move_x, opp_move_y);

    // Always swap strong openings
    if (STRONG_OPENINGS.count(key) > 0) {
        return true;
    }

    // Never swap weak openings
    if (WEAK_OPENINGS.count(key) > 0) {
        return false;
    }

    // For other moves, use distance from center heuristic
    int center = board_size / 2;
    int dist = std::abs(opp_move_x - center) + std::abs(opp_move_y - center);

    // Swap if within 2 tiles of center
    return dist <= 2;
}

double get_move_priority(const Board& board, std::pair<int, int> move, Colour colour) {
    int x = move.first;
    int y = move.second;
    double score = 0.0;

    // Bonus for extending existing connections
    auto neighbors = board.get_neighbors(x, y);
    for (const auto& [nx, ny] : neighbors) {
        Tile tile = board.get_tile(nx, ny);
        if (tile == (colour == Colour::RED ? Tile::RED : Tile::BLUE)) {
            score += 2.0;
        }
    }

    // Bonus for blocking opponent connections
    Colour opp = opposite_colour(colour);
    Tile opp_tile = (opp == Colour::RED ? Tile::RED : Tile::BLUE);
    for (const auto& [nx, ny] : neighbors) {
        if (board.get_tile(nx, ny) == opp_tile) {
            score += 1.5;
        }
    }

    // Bonus for center positions (strategic value)
    int center = BOARD_SIZE / 2;
    int dist_to_center = std::abs(x - center) + std::abs(y - center);
    score += std::max(0.0, (BOARD_SIZE - dist_to_center) / static_cast<double>(BOARD_SIZE));

    // Bonus for positions along main diagonal
    if (colour == Colour::RED) {
        // RED benefits from positions along y axis progression
        if (x == y) score += 0.5;
    } else {
        // BLUE benefits from positions along x axis progression
        if (x == y) score += 0.5;
    }

    return score;
}

std::pair<int, int> select_weighted_move(
    const std::vector<std::pair<int, int>>& empty_tiles,
    const Board& board,
    Colour colour
) {
    if (empty_tiles.empty()) {
        return {-1, -1};
    }

    if (empty_tiles.size() == 1) {
        return empty_tiles[0];
    }

    // For speed, only score a subset of moves
    int sample_size = std::min(8, static_cast<int>(empty_tiles.size()));

    static thread_local std::mt19937 gen(std::random_device{}());

    std::vector<std::pair<int, int>> candidates;
    if (static_cast<int>(empty_tiles.size()) <= sample_size) {
        candidates = empty_tiles;
    } else {
        // Random sample
        std::sample(empty_tiles.begin(), empty_tiles.end(),
                   std::back_inserter(candidates), sample_size, gen);
    }

    std::pair<int, int> best_move = candidates[0];
    double best_score = -std::numeric_limits<double>::infinity();

    std::uniform_real_distribution<> dis(0.0, 0.1);

    for (const auto& move : candidates) {
        double score = get_move_priority(board, move, colour);
        // Add small random noise to break ties
        score += dis(gen);

        if (score > best_score) {
            best_score = score;
            best_move = move;
        }
    }

    return best_move;
}

int shortest_path_distance(const Board& board, Colour colour) {
    constexpr int INF = BOARD_SIZE * BOARD_SIZE + 1;

    std::vector<std::vector<int>> dist(BOARD_SIZE, std::vector<int>(BOARD_SIZE, INF));

    // Priority queue: (distance, x, y)
    using PQElement = std::tuple<int, int, int>;
    std::priority_queue<PQElement, std::vector<PQElement>, std::greater<>> pq;

    Tile my_tile = (colour == Colour::RED ? Tile::RED : Tile::BLUE);
    Tile opp_tile = (colour == Colour::RED ? Tile::BLUE : Tile::RED);

    // Initialize starting positions
    if (colour == Colour::RED) {
        // Start from top row
        for (int j = 0; j < BOARD_SIZE; ++j) {
            Tile tile = board.get_tile(0, j);
            if (tile == opp_tile) continue;

            int cost = (tile == my_tile) ? 0 : 1;
            dist[0][j] = cost;
            pq.push({cost, 0, j});
        }
    } else {
        // Start from left column
        for (int i = 0; i < BOARD_SIZE; ++i) {
            Tile tile = board.get_tile(i, 0);
            if (tile == opp_tile) continue;

            int cost = (tile == my_tile) ? 0 : 1;
            dist[i][0] = cost;
            pq.push({cost, i, 0});
        }
    }

    while (!pq.empty()) {
        auto [d, x, y] = pq.top();
        pq.pop();

        if (d > dist[x][y]) continue;

        // Check goal
        if (colour == Colour::RED && x == BOARD_SIZE - 1) {
            return d;
        }
        if (colour == Colour::BLUE && y == BOARD_SIZE - 1) {
            return d;
        }

        // Explore neighbors
        auto neighbors = board.get_neighbors(x, y);
        for (const auto& [nx, ny] : neighbors) {
            Tile tile = board.get_tile(nx, ny);
            if (tile == opp_tile) continue;

            int edge_cost = (tile == my_tile) ? 0 : 1;
            int new_dist = d + edge_cost;

            if (new_dist < dist[nx][ny]) {
                dist[nx][ny] = new_dist;
                pq.push({new_dist, nx, ny});
            }
        }
    }

    // Find minimum distance to goal edge
    int min_dist = INF;
    if (colour == Colour::RED) {
        for (int j = 0; j < BOARD_SIZE; ++j) {
            min_dist = std::min(min_dist, dist[BOARD_SIZE - 1][j]);
        }
    } else {
        for (int i = 0; i < BOARD_SIZE; ++i) {
            min_dist = std::min(min_dist, dist[i][BOARD_SIZE - 1]);
        }
    }

    return min_dist;
}

double evaluate_position(const Board& board, Colour colour) {
    int my_dist = shortest_path_distance(board, colour);
    int opp_dist = shortest_path_distance(board, opposite_colour(colour));

    // Lower distance is better, so we want opp_dist - my_dist
    return opp_dist - my_dist;
}

} // namespace hex
